checkers/typeDefFirst_checker.go

```diff
@@ -0,0 +1,75 @@
+package checkers
+
+import (
+	"go/ast"
+	"go/token"
+	"strings"
+
+	"github.com/go-critic/go-critic/checkers/internal/astwalk"
+	"github.com/go-critic/go-critic/framework/linter"
+)
+
+func init() {
+	var info linter.CheckerInfo
+	info.Name = "typeDefFirst"
+	info.Tags = []string{"style", "experimental"}
+	info.Summary = "File-scoped checker, that requires type definition before its method definitions"
+	info.Before = `
+func (r rec) Method() {}
+type rec struct{}
+`
+	info.After = `
+type rec struct{}
+func (r rec) Method() {}
+`
+	collection.AddChecker(&info, func(ctx *linter.CheckerContext) linter.FileWalker {
+		return &typeDefFirstChecker{
+			ctx: ctx,
+		}
+	})
+}
+
+type typeDefFirstChecker struct {
+	astwalk.WalkHandler
+	ctx *linter.CheckerContext
+}
+
+func (t *typeDefFirstChecker) WalkFile(f *ast.File) {
+	typeUsageMap := make(map[string]bool)
+
+	if f.Decls == nil {
+		return
+	}
+
+	for _, declaration := range f.Decls {
+		switch decl := declaration.(type) {
+		case *ast.FuncDecl:
+			if decl.Recv != nil {
+				receiver := decl.Recv.List[0]
+				typeName := trimAsterisk(NodeToString(t.ctx.FileSet, receiver.Type))
+				typeUsageMap[typeName] = true
+			}
+
+		case *ast.GenDecl:
+			if decl.Tok == token.TYPE {
+				for _, spec := range decl.Specs {
+					if spec, ok := spec.(*ast.TypeSpec); ok {
```

Comment:
You can decrease the nesting even further:
```go
for _, spec := range decl.Specs {
	spec, ok := spec.(*ast.TypeSpec)
	if !ok {
		continue
	}
```
This idiom is quite common.

Reply:
I know, but `continue` in nested for it is a little harder to understand, that is why I asked

Reply:
You know the solution to that: don't nest a lot of fors. :)
It's always possible to define another method and call that, so you don't have too much control flow on the same level.
But I'm OK with the current level of nesting. AST requires a lot of things like this, unfortunately.